crates/e2e/src/subxt_client.rs

```diff
@@ -607,9 +571,162 @@ where
             _marker: Default::default(),
         }
     }
+    async fn instantiate_with_gas_margin<
+        Contract: Clone,
+        Args: Send + Sync + Encode + Clone,
+        R,
+    >(
+        &mut self,
+        contract_name: &str,
+        caller: &Keypair,
+        constructor: &mut CreateBuilderPartial<E, Contract, Args, R>,
+        value: E::Balance,
+        margin: Option<u64>,
+        storage_deposit_limit: Option<E::Balance>,
+    ) -> Result<InstantiationResult<E, Self::EventLog>, Self::Error> {
+        let code = self.contracts.load_code(contract_name);
+        let data = constructor_exec_input(constructor.clone());
+
+        let dry_run = self
+            .api
+            .instantiate_with_code_dry_run(
+                value,
+                storage_deposit_limit,
+                code.clone(),
+                data.clone(),
+                salt(),
+                caller,
+            )
+            .await;
```

Comment:
If the user wants the gas portion (or explicitly specifies a gas limit) we should just perform the dry-run from the builder itself via `bare_call_dry_run`